### PR TITLE
OHRM5X-1102: Develop performance - KPI screens

### PR DESCRIPTION
**Pull request by devishke-orange**
PR contains:
- OHRM5X-1102: [FE/BE] Performance - Configure - KPIs
    - Rebased for 5.1 (#1106) 
- OHRM5X-1104: [FE/BE] Performance - List My Trackers
    - FE rebased for 5.1 (#1109) . BE not yet developed.
- OHRM5X-1105: [FE/BE] Performance - Employee Trackers
    - FE rebased for 5.1 (#1108) . BE not yet developed.

### CODE REVIEW

File: src/plugins/orangehrmPerformancePlugin/config/routes.yaml [hunk 1]
```diff
@@ -1,4 +1,42 @@
...
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ GET, POST, DELETE ]
+  defaults:
+    _api: OrangeHRM\Performance\Api\KpiAPI
+
+##########################
+#        Screens
+##########################
+
+performance_myTrackers:
```
Comment (devishke-orange) [inline src/plugins/orangehrmPerformancePlugin/config/routes.yaml:19]:
```suggestion
performance_my_trackers:
```

File: installer/Migration/V5_1_0/Migration.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+use Doctrine\DBAL\Schema\ForeignKeyConstraint;
+use OrangeHRM\Installer\Util\V1\AbstractMigration;
+
+class Migration extends AbstractMigration
+{
+    public function up(): void
+    {
+        $this->getDataGroupHelper()->insertApiPermissions(__DIR__ . '/permission/api.yaml');
+
+        $this->getConnection()->executeStatement('ALTER TABLE ohrm_kpi CHANGE job_title_code job_title_code INT(13) NOT NULL');
```
Comment (devishke-orange) [inline installer/Migration/V5_1_0/Migration.php:31]: tried using `getSchemaHelper()->changeColumn()` but seems that for Integer type length does not get set

File: src/plugins/orangehrmPerformancePlugin/Api/KpiAPI.php [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            new ParamRule(
+                self::PARAMETER_MIN_RATING,
+                new Rule(Rules::INT_TYPE),
+                new Rule(Rules::BETWEEN, [0, 100])
+            ),
+            new ParamRule(
+                self::PARAMETER_MAX_RATING,
+                new Rule(Rules::INT_TYPE),
+                new Rule(Rules::BETWEEN, [0, 100])
+            ),
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmPerformancePlugin/Api/KpiAPI.php:210]: No. Service class throws exception.

File: src/client/src/orangehrmPerformancePlugin/pages/KpiSearch.vue [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+const kpiNormalizer = data => {
+  return data.map(item => {
+    return {
+      id: item.id,
+      title: item.title,
+      jobTitleName: item.jobTitle.name,
+      jobTitleId: item.jobTitle.id,
+      minRating: item.minRating,
+      maxRating: item.maxRating,
+      isDefault: item.isDefault ? 'Yes' : '',
```
Comment (RajithaKumara) [inline src/client/src/orangehrmPerformancePlugin/pages/KpiSearch.vue:104]: better to use boolean value from API

File: src/plugins/orangehrmPerformancePlugin/Api/KpiAPI.php [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        return new EndpointResourceResult(ArrayModel::class, $ids);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForDelete(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(CommonParams::PARAMETER_IDS)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Api/KpiAPI.php:266]: shall we introduce new rule as INT_ARRAY

File: src/plugins/orangehrmPerformancePlugin/Controller/KpiSaveController.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+use OrangeHRM\Framework\Http\Request;
+
+class KpiSaveController extends AbstractVueController
+{
+    /**
+     * @inheritDoc
+     */
+    public function preRender(Request $request): void
+    {
+        $id = $request->get('id');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Controller/KpiSaveController.php:34]:
```suggestion
        $id = $request->attribute->get('id');
```

File: src/plugins/orangehrmPerformancePlugin/Dao/KpiDao.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        $kpi = $this->getRepository(Kpi::class)->findOneBy(['id' => $id, 'deletedAt' => null]);
+        if ($kpi instanceof Kpi) {
+            return $kpi;
+        }
+        return null;
+    }
+
+    /**
+     * @return Kpi[]
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Dao/KpiDao.php:56]: update the doc comment

File: installer/Migration/V5_1_0/Migration.php (hunk 2 above)
Comment (RajithaKumara) [reply installer/Migration/V5_1_0/Migration.php:31]: check with Sharuka.

**Review by RajithaKumara — APPROVED** [review]
Mobile review. Devishke self review again. 
@Super-Chama hope you can approve FE code

File: installer/Migration/V5_1_0/Migration.php (hunk 2 above)
Comment (SharukaNP) [reply installer/Migration/V5_1_0/Migration.php:31]: Sorry I was not available yesterday will take a call today and discuss

File: installer/Util/AppSetupUtility.php [hunk 8]
```diff
@@ -60,6 +60,9 @@ class AppSetupUtility
             \OrangeHRM\Installer\Migration\V5_0_0_beta\Migration::class,
             \OrangeHRM\Installer\Migration\V5_0_0\Migration::class,
         ],
+        '5.1' => [
+            \OrangeHRM\Installer\Migration\V5_1_0\Migration::class,
+        ],
```
Comment (RajithaKumara) [inline installer/Util/AppSetupUtility.php:65]:
```suggestion
        '5.1' => \OrangeHRM\Installer\Migration\V5_1_0\Migration::class,
```

File: src/client/src/core/util/validation/rules.ts [hunk 9]
```diff
@@ -459,3 +459,51 @@ export const minValueShouldBeLowerThanMaxValue = (
     return parseFloat(resolvedMinValue) < parseFloat(value) || resolvedMessage;
   };
 };
+
+/**
+ * @param {string | function} maxValue
+ * @param {string|undefined} message
+ */
+
```
Comment (RajithaKumara) [inline src/client/src/core/util/validation/rules.ts:467]:
```suggestion
```

File: src/client/src/core/util/validation/rules.ts [hunk 10]
```diff
@@ -459,3 +459,51 @@ export const minValueShouldBeLowerThanMaxValue = (
...
+    return parseFloat(resolvedMaxValue) > parseFloat(value) || resolvedMessage;
+  };
+};
+
+/**
+ * @param {number} minValue
+ * @param {number} maxValue
+ * @param {string|undefined} message
+ */
+
```
Comment (RajithaKumara) [inline src/client/src/core/util/validation/rules.ts:491]:
```suggestion
```

File: src/client/src/orangehrmPerformancePlugin/pages/MyTracker.vue [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    const serializedFilter = computed(() => {
+      return {
+        sortField: sortField.value,
+        sortOrder: sortOrder.value,
+      };
+    });
+
+    const http = new APIService(
+      'https://990db1a5-734a-4bd6-a39b-12f29fbedfc3.mock.pstmn.io',
+      '/api/v2/leave/myTracker',
```
Comment (RajithaKumara) [inline src/client/src/orangehrmPerformancePlugin/pages/MyTracker.vue:77]:
```suggestion
      '/api/v2/performance/trackers',
```
:thinking:

File: src/client/src/orangehrmPerformancePlugin/pages/KpiSave.vue [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+export default {
+  name: 'KpiSave',
+  components: {
+    'oxd-switch-input': SwitchInput,
+    'jobtitle-dropdown': JobtitleDropdown,
+  },
+  setup() {
+    const http = new APIService(
+      window.appGlobal.baseUrl,
+      'api/v2/performance/kpi',
```
Comment (RajithaKumara) [inline src/client/src/orangehrmPerformancePlugin/pages/KpiSave.vue:128]: can use plural nouns for collection endpoint. 
https://swagger.io/resources/articles/best-practices-in-api-design/

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**